backupccl: add restore performance logging

pbardea opened the pull request.

pbardea committed:
backupccl: add restore performance logging

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/63560)
<!-- Reviewable:end -->

dt commented on pkg/ccl/backupccl/restore_data_processor.go:
```diff
@@ -188,6 +203,9 @@ func (rd *restoreDataProcessor) processRestoreSpanEntry(
+	timeOpeningIters := timeutil.Since(startTime)
+	rd.totalIterators += timeOpeningIters
+
 	batcher, err := bulk.MakeSSTBatcher(ctx, db, evalCtx.Settings,
```
I wonder if we should move this to creation of the processor, so that we can get the aggregate timing over the whole processor lifetime.

pbardea replied:
Moved this to the creation of the processor. I exposed the per-batch bulk op summary here. I still flush and reset on every entry though. I agree that it makes sense to move that out so that thee batcher can flush when it wants, but will address that in another PR since that change is more involved.

dt commented on pkg/ccl/backupccl/restore_data_processor.go:
```diff
@@ -251,19 +276,35 @@ func (rd *restoreDataProcessor) processRestoreSpanEntry(
 			return summary, errors.Wrapf(err, "adding to batch: %s -> %s", key, value.PrettyPrint())
 		}
+		timeAdding += timeutil.Since(beforeAdd)
+		lastAdd = timeutil.Now()
```
I wouldn't be surprised if these in the hot loop actually slow the loop considerably

pbardea replied:
Removed these. In restore2TB bumping `kv.bulk_ingest.batch_size` avoids size-triggeered flushes during AddMVCCKey, so the time spent in this loop is meaningfully different from the time based flushing below. I think measuring this difference is interesting to see if reading the iterator is taking a substantial amount of time that can be pipelined with the addition of the keys to the batcher.

adityamaru commented:
@pbardea not sure if you saw this but might be useful https://github.com/cockroachdb/cockroach/pull/63468

pbardea committed:
backupccl: create one batcher per restore data processor

Release note: None

pbardea committed:
bulk,kv: collect more timings around adding SSTables

Collect timings per receipient that we're waiting on during the
AddSSTable request.

Release note: None

pbardea commented:
RFAL
